nvmeof: add qos ability

gadididi opened the pull request:
# Describe what this PR does #

Adds QoS (Quality of Service) support for NVMe-oF CSI driver, allowing users to set IOPS and bandwidth limits on volumes at creation time and modify them dynamically at runtime.


## Is there anything that requires special attention ##

**New StorageClass parameters:**
- `rwIosPerSecond` - Read/Write IOPS limit
- `rwMbytesPerSecond` - Read/Write bandwidth limit (MB/s)
- `rMbytesPerSecond` - Read bandwidth limit (MB/s)
- `wMbytesPerSecond` - Write bandwidth limit (MB/s)

**Key decisions:**
- Uses NVMe-oF-specific parameter names (not RBD QoS keys) because the architectures are different: details [here](https://github.com/ceph/ceph-csi/issues/5694
)
- Implements `ControllerModifyVolume()` RPC for runtime QoS changes via VolumeAttributesClass (GA at K8s 1.34+ 
you can read more about it here: [volume-attributes-classes)](https://kubernetes.io/docs/concepts/storage/volume-attributes-classes/)
- Adds stub `ControllerExpandVolume()` with `EXPAND_VOLUME` capability to satisfy csi-resizer requirements (csi-resizer needs this capability to start, even when only using it for modify operations)
- Uses `GetControllerPublishSecretRef()` for credential retrieval in `ControllerModifyVolume()` as a temporary solution (TODO: implement `GetControllerExpandSecretRef()` with proper NVMeoF support in ceph-csi-config)

**Credential handling note:**
Since CSI spec doesn't define `csi.storage.k8s.io/controller-modify-secret-name` parameter and csi-resizer doesn't pass secrets to `ControllerModifyVolume()` according to [storageclass-secrets,](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html#storageclass-secrets) . The implementation retrieves credentials from ceph-csi-config ConfigMap using the existing `controllerPublishSecretRef` field. This follows the same pattern as other CSI operations.


**Example workflow:**

1. Create StorageClass with:
```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: csi-nvmeof-qos
provisioner: nvmeof.csi.ceph.com
parameters:
  clusterID: "cluster-1"
  pool: "mypool"
  # ... other params
```

2. Create default QoS

```yaml 
apiVersion: storage.k8s.io/v1
kind: VolumeAttributesClass
metadata:
  name: default
driverName: nvmeof.csi.ceph.com
parameters:
  rwIopsPerSecond: "5000"
```

3. Create PVC (gets 10000 IOPS, 100 MB/s):
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: my-pvc
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  storageClassName: csi-nvmeof-qos
  volumeAttributesClassName: default  # ← Start with default (5000 IOPS)
```

4. Later, define VolumeAttributesClass for runtime modification:
```yaml
apiVersion: storage.k8s.io/v1alpha1
kind: VolumeAttributesClass
metadata:
  name: high-performance
driverName: nvmeof.csi.ceph.com
parameters:
  nvmeof_rw_ios_per_second: "50000"
  nvmeof_rw_mbytes_per_second: "500"
```


5. Modify existing PVC (triggers ControllerModifyVolume):
```bash
kubectl patch pvc my-pvc -p '{"spec":{"volumeAttributesClassName":"high-performance"}}'
```

Result: Volume QoS updated to 50000 IOPS, 500 MB/s without recreation or downtime.

NOTE: 
- Adds csi-resizer sidecar to controller deployment according to this: [resizer](https://kubernetes.io/docs/concepts/storage/volume-attributes-classes/#resizer)

## Related issues ##

https://github.com/ceph/ceph-csi/issues/5694

## Future concerns ##
- Add proper NVMeoF struct to ceph-csi-config-map ClusterInfo with dedicated `controllerExpandSecretRef` field **OR** Add `controllerExpandSecretRef` for the existing `RBD` struct. (because NVMeoF is just a wrapper to RBD). 
- Implement `GetControllerExpandSecretRef()` to replace temporary use of `GetControllerPublishSecretRef()`

**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer
  guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release
  notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md)
  updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

gadididi committed:
nvmeof: add the call to qos in CreateVolume()

if qos params are provided, call to qos grpc after ns was created.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

gadididi committed:
nvmeof: adding od Qos grpc call

Qos for nvmeof namespace is added. allow the user limit the
namesapce capabilities.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

gadididi committed:
nvmeof: add ControllerModifyVolume() fucntion

the purpose of that function is to modify the qos for namesapce on the fly.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

gadididi committed:
nvmeof: add stub ControllerExpandVolume for csi-resizer compatibility

Add EXPAND_VOLUME capability and stub implementation to allow
csi-resizer to start and handle VolumeAttributesClass modifications.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

gadididi committed:
nvmeof: add modify_volume capability

modify_volume capability is added to nvmeof driver in order to call ControllerModifyVolume().

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

nixpanic commented on internal/nvmeof/qos.go:
```diff
@@ -0,0 +1,64 @@
+	WMbytesPerSecond  *uint64
+}
+
+// QoS parameter keys that will be shown in the storage class.
```
nit: "storage class" is Kubernetes specific, this would work with any container orchistrator. Maybe mention the "ModifyVolume" CSI procedure too?

gadididi replied:
fixed

nixpanic commented on internal/nvmeof/qos.go:
```diff
@@ -0,0 +1,64 @@
+
+// NVMeoFQosVolume holds the QoS parameters for an NVMe-oF volume.
+type NVMeoFQosVolume struct {
+	RwIosPerSecond    *uint64
```
Please comment each attribute with their non-abbreviated name.

gadididi replied:
fixed

nixpanic commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -351,6 +351,44 @@ func parseListeners(listenersJSON string) ([]nvmeof.ListenerDetails, error) {
 }
 
+// parseQoSParameters extracts and parses QoS parameters from the given map.
+func parseQoSParameters(params map[string]string) (*nvmeof.NVMeoFQosVolume, error) {
```
Can you add unit tests for this function?

gadididi replied:
I added

nixpanic commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -293,6 +293,112 @@ func (cs *Server) ControllerUnpublishVolume(
+	// Step 3: Get NVMe-oF metadata for the volume
+
+	// Since ControllerModifyVolume doesn't receive volume context and dont have option to take secrets
+	// because there is no "csi.storage.k8s.io/controller-modify-secret-name" field in the SC !,
```
`ControllerModifyVolume` can have secrets according to the specification: https://github.com/container-storage-interface/spec/blob/master/spec.md#controllermodifyvolume

Do you mean that the resizer sidecar does not pass the secret from Kubernetes to the CSI-driver? Is there an issue for that at https://github.com/kubernetes-csi/external-resizer , or can we provide a PR for it?

gadididi replied:
Yes, I saw it has that, but when I ran it the secret was `nil`, so I thought maybe I need to add into StorageClass 

```yml
  csi.storage.k8s.io/controller-modify-secret-name: csi-rbd-secret
  csi.storage.k8s.io/controller-modify-secret-namespace: default
```

like I have for **`ControllerPublishVolume()`** 

```yml
  csi.storage.k8s.io/controller-publish-secret-name: csi-rbd-secret
  csi.storage.k8s.io/controller-publish-secret-namespace: default
```
but what I see here:
https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html#storageclass-secrets
there is no for modify. maybe I am wrong, but I could not find any place that mentioned that..


So, I implemented like `ControllerUnPublishVolume()`  here:  https://github.com/ceph/ceph-csi/blob/f01ab138265cac37a9c6a9893daa7a0edb8663ec/internal/nvmeof/controller/controllerserver.go#L268

About the resizer, I am not sure how it should pass the secret. I will look at it and see if it is unimplemented\issue or something else..

nixpanic replied:
Secrets are indeed not fetched, they are always `nil`:

https://github.com/kubernetes-csi/external-resizer/blob/cba05cb7cc4d06d7c06ae95dc8941fdcb75bd67e/pkg/modifier/csi_modifier.go#L84-L89

I do not think it is possible to get secrets as part of the `ControllerModifyVolume` call, so these need to be resolved through an other way (like the new default credentials).

nixpanic replied:
maybe kubernetes-csi/external-resizer#544 is acceptable, would you be able to test with that container-image?

nixpanic commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -399,6 +399,21 @@ func (cs *Server) ControllerModifyVolume(
+	ctx context.Context,
+	req *csi.ControllerExpandVolumeRequest,
+) (*csi.ControllerExpandVolumeResponse, error) {
+	return &csi.ControllerExpandVolumeResponse{
```
instead of faking an action, can you return a gRPC UNIMPLEMENTED error?

gadididi replied:
fixed

nixpanic commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -399,6 +399,21 @@ func (cs *Server) ControllerModifyVolume(
+// For now it only updates the capacity in the response as NVMe-oF
+// this must be added because ControllerModifyVolume requires the sidecar csi-resizer. and
+// csi-resizer searches for the capacity ControllerServiceCapability_RPC_EXPAND_VOLUME.
+// In the future, if NVMe-oF gateway supports volume expansion, the logic must be added here.
```
This sounds like a shortcoming of the external-resizer sidecar. Only ModifyVolume should be allowed. But there seems to be a required check on volume resize: https://github.com/kubernetes-csi/external-resizer/blob/master/pkg/resizer/csi_resizer.go#L59

Madhu-1 replied:
@nixpanic i think we should open a tracker with the external-resizer as there could be drivers that support ModiyVolume but not Expand volume.

nixpanic replied:
created https://github.com/kubernetes-csi/external-resizer/pull/545 for that, but a **draft** for now as I have not tested it yet

nixpanic replied:
This was merged. The `ControllerExpandVolume` procedure can be kept until the external-resizer has a release. (Or maybe by that time nvmeof implements resizing too.)

gadididi replied:
@nixpanic , Aviv told me we should implement the `ControllerExpandVolume()` (resize nvmeof volume). So, in any case we need this function.

nixpanic replied:
yes,`ControllerExpandVolume` needs to be implemented too, but in a separate PR :smile:

gadididi committed:
rbd: add query if params have rbd qos vars

added because want to reuse the private function parseQosParams()

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

Madhu-1 commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -293,6 +294,78 @@ func (cs *Server) ControllerUnpublishVolume(
+	volumeID := req.GetVolumeId()
+	params := req.GetMutableParameters()
+
+	log.DebugLog(ctx, "ControllerModifyVolume: volumeID=%s, params=%+v", volumeID, params)
```
This should  be logged in the Server.go please add this new RPC to https://github.com/ceph/ceph-csi/blob/fd202c7cf1a9619d528351769c108a41b17480f5/internal/csi-common/utils.go#L210

gadididi replied:
Good catch!

Madhu-1 replied:
@gadididi Lets remove it as it gets logged as part of the RPC request it self?

gadididi replied:
fixed

Madhu-1 commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -293,6 +294,78 @@ func (cs *Server) ControllerUnpublishVolume(
+		// Should never reach here
+		log.ErrorLog(ctx, "Unexpected QoS parameter state")
+
+		return nil, status.Error(codes.Internal, "unexpected QoS parameter state")
```
invalid InvalidArgument?

gadididi replied:
my bad, will fix it

gadididi replied:
fixed

Madhu-1 commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -312,6 +385,17 @@ func validateCreateVolumeRequest(req *csi.CreateVolumeRequest) error {
 	if err != nil {
 		return fmt.Errorf("invalid listeners parameter: %w", err)
 	}
+	// Validate QoS parameters - cannot mix RBD and NVMe-oF QoS
+	hasRBDQoS := rbd.HasQoSParams(params)
+	nvmeofQoS, err := parseQoSParameters(params)
+	if err != nil {
+		return fmt.Errorf("invalid NVMe-oF QoS parameters: %w", err)
+	}
+	hasNVMeoFQoS := (nvmeofQoS != nil)
+
+	if hasRBDQoS && hasNVMeoFQoS {
+		return errors.New("cannot specify both RBD and NVMe-oF QoS parameters")
+	}
```
These should be MutableParameters not the parameters during the CreateVolume isnt it?

gadididi replied:
@Madhu-1, I saw there is field named  `MutableParameters` in `CreateVolumeRequest` struct. 
but I saw in rbd code it took the rbd QoS parameters from `req.GetParameters()`
here:
https://github.com/ceph/ceph-csi/blob/fd202c7cf1a9619d528351769c108a41b17480f5/internal/rbd/controllerserver.go#L437
then the flow comes to here:
https://github.com/ceph/ceph-csi/blob/fd202c7cf1a9619d528351769c108a41b17480f5/internal/rbd/controllerserver.go#L823-L830

So, if I understand correctly the rbd's QoS params are located in the SC. I wonder if put the nvmeof's QoS params in SC as `MutableParameters` , but I decided to do like rbd..

 what do you think?

nixpanic replied:
The RBD implementation does not use `ControllerModifyVolume` (yet). For `ControllerModifyVolume` the `mutable_parameters` in the StorageClass should be used (relatively new addition). Prevent using `parameters` in the StorageClass for QoS.

Madhu-1 replied:
yes we should move to the mutable as per the CSI spec. lets open a tracker for it for nbd QOS as well.

gadididi replied:
Ok, so now I am going to change just `nvmeofQoS, err := parseQoSParameters(mutableParams)`
and leave this line `hasRBDQoS := rbd.HasQoSParams(params)` as is ( I add TODO, once RBD will implement the ControllerModifyVolume, we can change )

I read how it works:
**There is no section in SC that called `mutable_parameters` like `parameters`**

there will be default VolumeAttributesClass . and on the fly the user can modify with other VACs:

### example
1. StorageClass (no QoS):
```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: csi-nvmeof-sc
provisioner: nvmeof.csi.ceph.com
parameters:
  pool: mypool
  clusterID: "..."
  # No QoS here!
```
2. VolumeAttributesClass "default" (5000 IOPS):
```yaml
apiVersion: storage.k8s.io/v1
kind: VolumeAttributesClass
metadata:
  name: default
driverName: nvmeof.csi.ceph.com
parameters:
  rwIopsPerSecond: "5000"
```

**create SC and default VAC**

3. PVC (uses "default"):
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: my-pvc
spec:
  storageClassName: csi-nvmeof-sc
  volumeAttributesClassName: default  # ← Start with default (5000 IOPS)
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
```
**Volume is created with 5000 IOPS** 

**Later: Modify to High Performance**
5. Create "high-performance" VolumeAttributesClass (10000 IOPS):
```yaml
apiVersion: storage.k8s.io/v1
kind: VolumeAttributesClass
metadata:
  name: high-performance
driverName: nvmeof.csi.ceph.com
parameters:
  rwIopsPerSecond: "10000"
  rwMbytesPerSecond: "100"
```
6. Patch the PVC to use "high-performance":

```bash
bashkubectl patch pvc my-pvc -p '{"spec":{"volumeAttributesClassName":"high-performance"}}'
```

Once RBD controller will implement `ControllerModifyVolume()` some PVC can have VAC of default rbd QoS , and some can have VAC of default nvmeof QoS (what the user wants) , and some with nothing (unlimited) and the `req.GetMutableParameters()` will return `nil`  .

Madhu-1 replied:
Yes thats correct expectation from me as well.

Madhu-1 commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -351,6 +435,131 @@ func parseListeners(listenersJSON string) ([]nvmeof.ListenerDetails, error) {
+		return nil, err
+	}
+
+	if !hasAnyQoS {
```
we dont need this variable right? we can simply return qos which will be nil if nothing is set

gadididi replied:
🤔 
maybe I am missing something, but:
```go
qos := &nvmeof.NVMeoFQosVolume{}  // ← This creates a NON-NIL pointer to an empty struct (that contains 4 nil pointers)
```
so, I can remove the `hasAnyQoS` , but I still need to add to this function:
```go
	// Return nil if no QoS parameters were set
	if qos.RwIosPerSecond == nil && qos.RwMbytesPerSecond == nil &&
		qos.RMbytesPerSecond == nil && qos.WMbytesPerSecond == nil {
		return nil, nil
	}
```

nixpanic replied:
I think it is simpler with the `hasAnyQoS` bool. You need to allocate a `qos` struct so that the values can be set in it.

Madhu-1 commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -351,6 +435,131 @@ func parseListeners(listenersJSON string) ([]nvmeof.ListenerDetails, error) {
+
+		return nil, status.Errorf(codes.Aborted, "subsystem %s operation already in progress", nvmeofData.SubsystemNQN)
+	}
+	defer cs.subsystemLocks.Release(nvmeofData.SubsystemNQN)
```
would it make sense to acquire lock and than connect to sub system?

nixpanic replied:
Maybe the subsystem lock is not even needed here?

Madhu-1 replied:
lets remove it if its not required

gadididi replied:
I checked, we can remove it

Madhu-1 commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -351,6 +435,131 @@ func parseListeners(listenersJSON string) ([]nvmeof.ListenerDetails, error) {
+	err = gateway.SetQoSLimitsForNamespace(ctx, nvmeofData.SubsystemNQN, nvmeofData.NamespaceID, *qos)
+	if err != nil {
+		// Check if error is EEXIST (RBD QoS already set)
+		if errors.Is(err, os.ErrExist) {
```
can we use cephCSI specific error (declare it in error.go) instead of using os.ExxExist?

gadididi replied:
addressed

Madhu-1 commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -351,6 +435,131 @@ func parseListeners(listenersJSON string) ([]nvmeof.ListenerDetails, error) {
+		if errors.Is(err, os.ErrExist) {
+			log.ErrorLog(ctx, "RBD QoS already configured on volume")
+
+			return nil, status.Error(codes.InvalidArgument,
```
this need to be success when it already configured with same values if not it need to be changed right?

gadididi replied:
@Madhu-1 , the "already exists" means there is RBD QoS in that namespace(=rbd image), and if the user tries also set QoS for the nvmeof namespace (that effects on the spdk, not directly on the rbd image) it should fail.

the logic is:
```md
1. Parse mutable_parameters
2. Check if contains RBD QoS params? 
   - YES: 
     - Check if ALSO contains NVMe-oF QoS params?
       - YES: Return INVALID_ARGUMENT (can't have both)
       - NO: Call RBD ControllerModifyVolume, return
   - NO: Continue with NVMe-oF QoS
3. Apply NVMe-oF QoS via gateway
4. If gateway returns EEXIST: Return INVALID_ARGUMENT (RBD QoS already set)
5. If other gateway error: Return INTERNAL
```

the current RBD controller does not have   `ControllerModifyVolume` implementation. later in nvmeof `ControllerModifyVolume` we will call to rbd `ControllerModifyVolume` if only rbd QoS params are provided.

Madhu-1 replied:
got it, i assume if we reissue the call again for the NVMe-of, the call will not fail.

gadididi replied:
yes, correct

Madhu-1 commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -465,6 +673,20 @@ func (cs *Server) createNVMeoFResources(
 		},
 	}
+	// extract Qos parameters if any
+	nvmeofQoS, err := parseQoSParameters(params)
```
this should be MutableParameters not the regular one

gadididi replied:
same topic\question as here: https://github.com/ceph/ceph-csi/pull/5714/files/359a34bb20db086c5e6a19200c6e710a24c69655#r2509807057

Madhu-1 replied:
for Nvme lets do it right. use mutable parameters only

gadididi replied:
fixed

gadididi committed:
nvmeof: add errors.go for nvmeof driver

init that file. in the future add more nvmeof errors to this file.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

gadididi committed:
util: add ControllerModifyVolume to request ID extraction

add ControllerModifyVolume to request ID extraction for proper log correlation.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

nixpanic commented on internal/nvmeof/qos.go:
```diff
@@ -0,0 +1,66 @@
+	RwIosPerSecond    = "nvmeof_rw_ios_per_second"
+	RwMbytesPerSecond = "nvmeof_rw_mbytes_per_second"
+	RMbytesPerSecond  = "nvmeof_r_mbytes_per_second"
+	WMbytesPerSecond  = "nvmeof_w_mbytes_per_second"
```
note that we do not have any parameters in the StorageClass that contain under_scores. Please use camelCase format for parameter keys. There is no need for the `nvmeof` prefix either, these parameters should only be documented for the NVMe-oF controller/provisioner.

Madhu-1 replied:
yes lets use CephCSI specifc keys for user facing configuration and translate it internall to keep the keys standard

gadididi replied:
Ok , I will use camelCase format.
FYI, these params will not be shown in the SC anymore, only in VAC yaml file.
I added an example in this PR here:
https://github.com/ceph/ceph-csi/pull/5714#discussion_r2517646297
 
 I think if we have 2 different types of VACs , for rbd and for nvmeof (once the rbd will implement the `modify`). 
 the VAC has field `driverName: nvmeof.csi.ceph.com`, it means only PVCs that created by nvmeof driver can be affected by this VAC. **BUT** I can create VAC for "nvmeof PVC" that has rbd qos params. 
 So I must somehow distinguish between these vars.
 I can remove the nvmeof prefix, to be e.g. `RwIosPerSecond` but it is very general, is not? 
 in this VAC I need to know what params are for nvmeof and what for rbd.
 with the nvmeof prefix it will be clearner IMO , what do you think? @nixpanic , @Madhu-1

Madhu-1 replied:
@gadididi lets take a step back, do we need to support RBD QOS for the rbd images created using nvmeof csi driver? what are the advantages of it? if there is no value we should not support it, just support QOS using nvmeof only

gadididi replied:
@Madhu-1, I did not see specific requirement for that. but I think if there is already implementation of QoS in the RBD controller, and there is an option to set rbd QoS on nvmeof volumes (via rbd cli on the rbd image), maybe is it good to have this option here (in nvmeof controller too)

UPDATE:
I was wrong about the RBD QoS 
I asked my manager Aviv, and he said, it is not recommended to have both together (as I said before) **but also it is not recommended to allow to the user to set QoS on rbd , when the user uses nvmeof volumes**. the purpose of the nvmeof is expose higher level of setting 

So, what I am going to do:


```md
1. Parse mutable_parameters
2. Check if contains RBD QoS params?
   - YES: Return INVALID_ARGUMENT (RBD QoS not supported for NVMe-oF volumes)
   - NO: Continue
3. Check if contains NVMe-oF QoS params?
   - NO: Return SUCCESS (no-op, nothing to modify)
   - YES: Continue
4. Apply NVMe-oF QoS via gateway
5. If gateway error EXIST: means rbd QoS was set (not recommended for the user , but he did it ) return ErrorExistst
6. if gateway return another error :  Return INTERNAL
7. else Return SUCCESS
```

what do you think?

nixpanic replied:
>  I can remove the nvmeof prefix, to be e.g. RwIosPerSecond but it is very general, is not?

Yes, the name of the parameter is general. But the StorageClass is for NVMe-oF, so it is still specific enough that way.

Madhu-1 replied:
sounds good and it also keep very minimal and exactly what we need to have

Madhu-1 commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -293,6 +294,78 @@ func (cs *Server) ControllerUnpublishVolume(
+		// No QoS parameters - nothing to do
+		log.DebugLog(ctx, "No QoS parameters provided for volume %s", volumeID)
+
+		return &csi.ControllerModifyVolumeResponse{}, nil
```
Question:- If the values are unset or removed from the VAC, dont we need to remove these QOS from the volume as well. In other words how a user can remove QOS?

gadididi replied:
@Madhu-1 , Removing `volumeAttributesClassName` doesn't trigger `ControllerModifyVolume()`
So, how to remove QoS (means set to unlimited all params), 2 options:

1. create removal_qos VAC with empty params. once I get it (in the ControllerModifyVolume()) by patching the new removal_qos VAC , I see no params at all - means set to unlimited.

2.  create removal_qos VAC with 0 params to all 4 fields. once I get it (in the ControllerModifyVolume()) by patching the new removal_qos VAC , it calls as always to the grpc set_qos .

currently code support option 2. if I get empty VAC (no params) it finishes with no error
IMO option 2 is better. what do you think?

Madhu-1 replied:
>Removing volumeAttributesClassName doesn't trigger ControllerModifyVolume()

Is this something supported? we need to test with kubernetes 1.34, If yes yes option 2 make sense to me.  we need to document these behaviour to set the right expectations.

nixpanic replied:
> create removal_qos VAC with 0 params to all 4 fields

Call the VAC **unlimited** or similar (in the documentation or examples), makes it easier for users to understand.

Madhu-1 commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -351,6 +435,131 @@ func parseListeners(listenersJSON string) ([]nvmeof.ListenerDetails, error) {
+		return nil
+	}
+
+	if err := parseParam(nvmeof.RwIosPerSecond, nvmeof.RwIosPerSecond, &qos.RwIosPerSecond); err != nil {
```
what will be the behavior if `0` is set for these keys. is it considered as infinite or disabled?

gadididi replied:
it is considered as infinite. but "disable QoS" means remove the limitation, so I think it is same terms

Madhu-1 commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -351,6 +435,131 @@ func parseListeners(listenersJSON string) ([]nvmeof.ListenerDetails, error) {
+	if err != nil {
+		log.ErrorLog(ctx, "Failed to get NVMe-oF metadata: %v", err)
+
+		return nil, status.Errorf(codes.NotFound, "NVMe-oF metadata not found: %v", err)
```
this is not always NotFound error right? it could be internal server error as well

gadididi replied:
fixed ( I added some errors in `errors.go` file ) and handling it in `getNVMeoFMetadata()` function

nixpanic commented on internal/nvmeof/nvmeof.go:
```diff
@@ -167,6 +168,43 @@ func (gw *GatewayRpcClient) DeleteNamespace(ctx context.Context, subsystemNQN st
+		RwIosPerSecond:    qos.RwIosPerSecond,
+		RwMbytesPerSecond: qos.RwMbytesPerSecond,
+		RMbytesPerSecond:  qos.RMbytesPerSecond,
+		WMbytesPerSecond:  qos.WMbytesPerSecond,
```
are all these parameters required? if a parameter is not set in `qos`, there should not be a default 0 configured. Instead, the configuration should not overwrite the missing parameter, but not set that at all.

gadididi replied:
e.g. if `qos.RwIosPerSecond` was not provided in the VAC, it means the value is nil (not 0)  
When you pass nil, the gateway won't receive that field at all (every field here are optional), so it won't overwrite existing values.

gadididi committed:
nvmeof: change getNVMeoFMetadata() function returns

make getNVMeoFMetadata() returns nvmeof error codes.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

gadididi committed:
nvmeof: add unit tests for parseQoSParameters

add unit tests with multiple cases.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

Madhu-1 commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -293,6 +294,54 @@ func (cs *Server) ControllerUnpublishVolume(
+	}
+	nvmeofQoS, err := parseQoSParameters(params)
+	if err != nil {
+		log.ErrorLog(ctx, "There are QOS params and it failed to parse QoS parameters: %v", err)
```
```suggestion
log.ErrorLog(ctx, "failed to parse NVMe-oF QoS parameters: %v", err)
```

gadididi replied:
fixed

Madhu-1 commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -312,6 +361,22 @@ func validateCreateVolumeRequest(req *csi.CreateVolumeRequest) error {
+	// TODO: currently the RBD QoS params are located in the StorageClass parameters (params),
+	// once RBD will implement the ControllerModifyVolume
+	// and we move the QoS params to mutable_parameters, we should change here to
+	// hasRBDQoS := rbd.HasQoSParams(mutableParams).
```
Lets remove this TODO?

lets add similar check for the mutableParams as well to fail the Create it self?

gadididi replied:
fixed, checks both mutableParams and Params

Madhu-1 reviewed (commented):
small nits, LGTM

nixpanic commented:
@mergifyio queue

gadididi commented:
@mergifyio requeue

gadididi commented:
@mergifyio queue

gadididi commented:
@mergifyio requeue

nixpanic commented:
@mergifyio queue
